### PR TITLE
Improve handling of **kwargs in FromParams

### LINKED ISSUE #4592: **kwargs handling is not ideal when interfacing with non-FromParams code
I am encountering

```
 File "/opt/conda/lib/python3.7/site-packages/torch_geometric/nn/conv/rgcn_conv.py", line 96, in __init__
    super(RGCNConv, self).__init__(aggr=aggr, node_dim=0, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'serialization_dir'
```
To my understanding, this is because **extras is passed to **kwargs, so the `__init__` of `MessagePassing` in pytorch_geometric reported an error.

Is there a way to prevent **extra to pass to a certain `Registrable`?  In my case, `Graph2GraphEncoder`.
Thanks!

Here is my code for `Graph2GraphEncoder`.
```
from typing import Callable

import torch
import torch_geometric ## nn __init__ has from .conv import * can use directly
from overrides import overrides

from allennlp.common import Registrable
#from torch_geometric.nn import GATConv, RGCNConv, CGConv

"""
An `Graph2GraphEncoder` is known as `GraphConvolutionLayer`
that takes some graph input and return its updated version.
For the most part we just use
[PyTorch Geometric GraphConvs](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html).
Here we provide a thin wrapper to allow registering them and instantiating them `from_params`.
The available options are
* ["gat"](link_to_gat)
* ["rgcn"](link_to_rgcn)
"""


class Graph2GraphEncoder(torch.nn.Module, Registrable):
    """
    A `Graph2GraphEncoder` is a `Module` that
    takes a graph and returs its updated version
    
    the API to outside world are `self.in_channels`, `self.out_channels`
    
    Input shape :
        `(SparseBatch, SparseGraphBatch)`;
    output shape:
        `(SparseBatch, SparseGraphBatch)`.
    """
    
    def get_input_dim(self) -> int: # not used
        """
        Returns the dimension of the vector input for each element in the sequence input
        to a `Seq2VecEncoder`.
        """
        return self.in_channels
        
    def get_output_dim(self) -> int: # not used
        """
        Returns the dimension of the final vector output by this `Seq2VecEncoder`.  This is `not`
        the shape of the returned tensor, but the last element of that shape.
        """
        return self.out_channels
        #raise NotImplementedError


class _Graph2GraphEncoderLambda(torch.nn.Module):
    """
    Wrapper around non PyTorch, lambda based Graph2GraphEncoders
    to display them as modules whenever printing model.
    """

    def __init__(self, func: Callable[[torch.Tensor], torch.Tensor], name: str):
        super().__init__()
        self._name = name
        self._func = func

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self._func(x)

    @overrides
    def _get_name(self):
        return self._name


# There are no classes to decorate, so we hack these into Registrable._registry.
# If you want to instantiate it, you can do like this:
# <T:Registrable>.by_name('relu')()
Registrable._registry[Graph2GraphEncoder] = {
    "gat": (torch_geometric.nn.conv.GATConv, None),
    "rgcn": (torch_geometric.nn.conv.RGCNConv, None), 
    # "linear": (lambda: _Graph2GraphEncoderLambda(lambda x: x, "Linear"), None),  # type: ignore
}

```

### LINKED ISSUE #4614: FromParams doesn't handle kwargs on base class
## Checklist

- [x] I have verified that the issue exists against the `master` branch of AllenNLP.
- [x] I have read the relevant section in the [contribution guide](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md#bug-fixes-and-new-features) on reporting bugs.
- [x] I have checked the [issues list](https://github.com/allenai/allennlp/issues) for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/allenai/allennlp/pulls) for existing proposed fixes.
- [x] I have checked the [CHANGELOG](https://github.com/allenai/allennlp/blob/master/CHANGELOG.md) and the [commit log](https://github.com/allenai/allennlp/commits/master) to find out if the bug was already fixed in the master branch.
- [x] I have included in the "Description" section below a traceback from any exceptions related to this bug.
- [x] I have included in the "Related issues or possible duplicates" section beloew all related issues and possible duplicate issues (If there are none, check this box anyway).
- [x] I have included in the "Environment" section below the name of the operating system and Python version that I was using when I discovered this bug.
- [x] I have included in the "Environment" section below the output of `pip freeze`.
- [x] I have included in the "Steps to reproduce" section below a minimally reproducible example.

## Description

If your create class that inherits from `FromParams`, and this class takes `**kwargs` in its constructor, then calling the `.from_params` method on this class will fail. See "Steps to reproduce" below.

## Related issues or possible duplicates

- This issue is similar: https://github.com/allenai/allennlp/issues/4592

## Environment

OS: OS X

Python version: Python 3.7

<details>
<summary><b>Output of <code>pip freeze</code>:</b></summary>
<p>

```
-e git+git@github.com:allenai/allennlp.git@e840a589afc4bfdac0165a8650145259a7603807#egg=allennlp
-e git+git@github.com:allenai/allennlp-models.git@2e5444997f416c112c0c6e02cc33175c24982049#egg=allennlp_models
apipkg==1.5
appdirs==1.4.4
appnope==0.1.0
attrs==20.1.0
backcall==0.2.0
black==20.8b1
bleach==3.1.5
blis==0.4.1
boto3==1.14.51
botocore==1.17.51
catalogue==1.0.0
certifi==2020.6.20
chardet==3.0.4
click==7.1.2
codecov==2.1.9
colorama==0.4.3
conllu==4.0
coverage==5.2.1
cycler==0.10.0
cymem==2.0.3
decorator==4.4.2
docutils==0.15
execnet==1.7.1
filelock==3.0.12
flake8==3.8.3
flaky==3.7.0
ftfy==5.8
future==0.18.2
greenlet==0.4.16
h5py==2.10.0
idna==2.10
importlib-metadata==1.7.0
iniconfig==1.0.1
ipython==7.18.0
ipython-genutils==0.2.0
jedi==0.17.2
Jinja2==2.11.2
jmespath==0.10.0
joblib==0.16.0
jsonnet==0.16.0
jsonpickle==1.4.1
keyring==21.3.1
kiwisolver==1.2.0
livereload==2.6.3
lunr==0.5.8
Markdown==3.2.2
markdown-include==0.5.1
MarkupSafe==1.1.1
matplotlib==3.3.1
mccabe==0.6.1
mkdocs==1.1.2
mkdocs-material==5.5.11
mkdocs-material-extensions==1.0
more-itertools==8.5.0
msgpack==0.5.6
murmurhash==1.0.2
mypy==0.782
mypy-extensions==0.4.3
neovim==0.2.6
nltk==3.5
nr.collections==0.0.1
nr.databind.core==0.0.16
nr.databind.json==0.0.13
nr.interface==0.0.3
nr.metaclass==0.0.5
nr.parsing.date==0.3.0
nr.pylang.utils==0.0.3
nr.stream==0.0.4
nr.utils.re==0.1.0
numpy==1.19.1
overrides==3.1.0
packaging==20.4
parso==0.7.1
pathspec==0.8.0
pathtools==0.1.2
pexpect==4.8.0
pickleshare==0.7.5
Pillow==7.2.0
pkginfo==1.5.0.1
plac==1.1.3
pluggy==0.13.1
preshed==3.0.2
prompt-toolkit==3.0.6
protobuf==3.13.0
ptyprocess==0.6.0
py==1.9.0
py-cpuinfo==7.0.0
py-rouge==1.1
pycodestyle==2.6.0
pydoc-markdown @ git+https://github.com/NiklasRosenstein/pydoc-markdown.git@f0bf8af1db4f11581c19d206d4ed1ab34b4854c1
pydocstyle==5.1.0
pyflakes==2.2.0
Pygments==2.6.1
pymdown-extensions==8.0
pyparsing==2.4.7
pytest==6.0.1
pytest-benchmark==3.2.3
pytest-cov==2.10.1
pytest-forked==1.3.0
pytest-xdist==2.1.0
python-dateutil==2.8.1
PyYAML==5.3.1
readme-renderer==26.0
regex==2020.7.14
requests==2.24.0
requests-toolbelt==0.9.1
responses==0.11.0
rfc3986==1.4.0
ruamel.yaml==0.16.10
ruamel.yaml.clib==0.2.0
s3transfer==0.3.3
sacremoses==0.0.43
scikit-learn==0.23.2
scipy==1.5.2
sentencepiece==0.1.91
six==1.15.0
snowballstemmer==2.0.0
spacy==2.3.2
srsly==1.0.2
tensorboardX==2.1
thinc==7.4.1
threadpoolctl==2.1.0
tokenizers==0.8.1rc1
toml==0.10.1
torch==1.6.0
tornado==6.0.4
tqdm==4.48.2
traitlets==4.3.3
transformers==3.0.2
twine==3.2.0
typed-ast==1.4.1
typing-extensions==3.7.4.3
urllib3==1.25.10
wasabi==0.8.0
watchdog==0.10.3
wcwidth==0.2.5
webencodings==0.5.1
word2number==1.1
zipp==3.1.0
```

</p>
</details>

## Steps to reproduce

<details>
<summary><b>Example source:</b></summary>
<p>

```python
from allennlp.common.from_params import FromParams, Params

class Foo(FromParams):
    def __init__(self, a: int, b: str = None, **kwargs) -> None:
        self.a = a
        self.b = b

Foo.from_params(Params({"a": 2, "b": "hi"}))
```

</p>
</details>

<details>
<summary><b>Python traceback:</b></summary>
<p>

```
Traceback (most recent call last):
  File "tmp.py", line 10, in <module>
    Foo.from_params(Params({"a": 2, "b": "hi"}))
  File "/Users/evanw/AllenAI/allennlp/allennlp/common/from_params.py", line 610, in from_params
    kwargs = create_kwargs(constructor_to_inspect, cls, params, **extras)
  File "/Users/evanw/AllenAI/allennlp/allennlp/common/from_params.py", line 163, in create_kwargs
    parameters = infer_params(cls, constructor)
  File "/Users/evanw/AllenAI/allennlp/allennlp/common/from_params.py", line 140, in infer_params
    super_parameters = infer_params(super_class)
  File "/Users/evanw/AllenAI/allennlp/allennlp/common/from_params.py", line 139, in infer_params
    raise RuntimeError("found a kwargs parameter with no inspectable super class")
RuntimeError: found a kwargs parameter with no inspectable super class
```

</p>
</details>

I would expect `Foo.from_params(Params({"a": 2, "b": "hi"}))` to work just fine. And if additional keyword argument were passed to `.from_params`, I would expect those to be passed on to `Foo`'s constructor as is.

### PR DESCRIPTION
**Pull request by matt-gardner**
Fixes #4614, and fixes #4592.

### CODE REVIEW

File: tests/common/from_params_test.py [hunk 1]
```diff
@@ -862,3 +861,93 @@ def __init__(self, nested_class: Optional[Union[str, NestedClass]] = None):
...
+            def __init__(self, c: str = None) -> None:
+                self.c = c
+
+        class Foo(Bar):
+            def __init__(self, a: int, b: str = None, **kwargs) -> None:
+                super().__init__(**kwargs)
+                self.a = a
+                self.b = b
+                for key, value in kwargs.items():
+                    setattr(self, key, value)
```
Comment (epwalsh) [inline tests/common/from_params_test.py:894]: Do we need these two lines? I think this is redundant given that we're passing kwargs to `super().__init__()`.

File: tests/common/from_params_test.py [hunk 2]
```diff
@@ -862,3 +862,93 @@ def __init__(self, nested_class: Optional[Union[str, NestedClass]] = None):
...
+
+        foo = Foo.from_params(Params({"a": 2, "b": "hi"}))
+        assert foo.a == 2
+        assert foo.b == "hi"
+
+        foo = Foo.from_params(Params({"a": 2, "b": "hi", "c": {"2": "3"}}))
+        assert foo.a == 2
+        assert foo.b == "hi"
+        assert foo.c == {"2": "3"}
```
Comment (epwalsh) [inline tests/common/from_params_test.py:881]: I think we're missing a test case where we have a `FromParams`  base class like `Foo` here that takes `**kwargs`, and then a subclass of `Foo` that takes `**kwargs` and just passes them on to its super class constructor:

```suggestion
        assert foo.c == {"2": "3"}
        
        class Bar(Foo):
            def __init__(self, a: int,  b: str, d: int, **kwargs) -> None:
                super().__init__(a, b=b, **kwargs)
                self.d = d
                
        bar = Bar.from_params(Params({"a": 2, "b": "hi", "c": {"2": "3"}, "d": 0}))
        assert bar.a == 2
        assert bar.b == "hi"
        assert bar.c == {"2": "3"}
        assert bar.d == 0
```

File: tests/common/from_params_test.py (hunk 1 above)
Comment (matt-gardner) [reply tests/common/from_params_test.py:894]: Yep, you're right, that was a copy-paste error.

File: tests/common/from_params_test.py (hunk 2 above)
Comment (matt-gardner) [reply tests/common/from_params_test.py:881]: Hmm, yeah, I didn't think to check that.  Looks like it already passes, though.

**Pull request merged**